File: go/pkg/protocol/transaction.go
package protocol

import "time"

type IsolationLevel string

const (
	IsolationReadUncommitted IsolationLevel = "read_uncommitted"
	IsolationReadCommitted   IsolationLevel = "read_committed"
	IsolationRepeatableRead  IsolationLevel = "repeatable_read"
	IsolationSerializable    IsolationLevel = "serializable"
)

type TransactionStatus string

const (
	TxnActive     TransactionStatus = "active"
	TxnCommitted  TransactionStatus = "committed"
	TxnRolledBack TransactionStatus = "rolled_back"
)

type TransactionInfo struct {
	ID             string            `json:"id" msgpack:"id"`
	ConnectionID   string            `json:"connection_id" msgpack:"connection_id"`
	Status         TransactionStatus `json:"status" msgpack:"status"`
	StartedAt      time.Time         `json:"started_at" msgpack:"started_at"`
	IsolationLevel IsolationLevel    `json:"isolation_level,omitempty" msgpack:"isolation_level,omitempty"`
}

type BeginTransactionRequest struct {
	ConnectionID   string         `json:"connection_id" msgpack:"connection_id"`
	IsolationLevel IsolationLevel `json:"isolation_level,omitempty" msgpack:"isolation_level,omitempty"`
}

File: rust/crates/protocol/src/transaction.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::common::ConnectionId;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransactionInfo {
    pub id: uuid::Uuid,
    pub connection_id: ConnectionId,
    pub status: TransactionStatus,
    pub started_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub isolation_level: Option<IsolationLevel>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransactionStatus {
    Active,
    Committed,
    RolledBack,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BeginTransactionRequest {
    #[serde(default = "uuid_default")]
    pub connection_id: ConnectionId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub isolation_level: Option<IsolationLevel>,
}

/// 客户端 body 不带 `connection_id`（handler 从 URL 路径补）时的占位值。
fn uuid_default() -> uuid::Uuid {
    uuid::Uuid::nil()
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum IsolationLevel {
    ReadUncommitted,
    #[default]
    ReadCommitted,
    RepeatableRead,
    Serializable,
}
